harness/determined/pytorch/_lightning/adapter.py

```diff
@@ -0,0 +1,184 @@
+from determined_common import check
+
+TorchData = Union[Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor]
+
+
+def check_compat(lm: pl.LightningModule) -> None:
+    prefix = "Unsupported usage in PLAdapter: "
+    unsupported_members = {"training_step_end", "validation_step_end", "manual_backward"}
+
+    members = inspect.getmembers(lm, predicate=inspect.ismethod)
+    overridden_members = set(
+        map(lambda m: m[0], filter(lambda m: is_overridden(m[0], lm), members))
+    )
+
+    matches = unsupported_members & overridden_members
+    if len(matches) > 0:
+        raise UnsupportedException(prefix + f"{matches}")
+
+    for member in overridden_members:
+        if has_param(getattr(lm, member), "dataloader_idx"):
+            raise UnsupportedException(
+                prefix
+                + f'multiple dataloaders and `dataloader_idx` are not supported in "{member}"'
+            )
+
+    if has_param(lm.training_step, "hiddens", 4):
+        raise UnsupportedException(prefix + '`hiddens` argument in "training_step"')
+
+    if lm.trainer is not None:
+        raise UnsupportedException(prefix + "Lightning Trainer")
+
+
+class PLAdapter(PyTorchTrial):
+    context: PyTorchTrialContext
+    lm: pl.LightningModule
+
+    def __init__(self, context: PyTorchTrialContext, lightning_module: pl.LightningModule):
+        super().__init__(context)
+        check_compat(lightning_module)
+        context.wrap_model(lightning_module)
+        self.lm = lightning_module
+        self.context = context
+
+        optimizers, lr_schedulers = self.setup_optimizers_schedulers(context)
+        self.optimizers = optimizers
+
+        # set lightning_module properties
+        self.lm.use_ddp = False
+        self.lm.use_ddp2 = False
+        self.lm.use_dp = False
+        self.lm.use_tpu = False
+        type(self.lm).local_rank = self.context.distributed.get_local_rank()
+        type(self.lm).global_rank = self.context.distributed.get_rank()
+        self.lm.use_amp = self.context._use_amp
+        self.lm.to(self.context.device)
+
+    def build_callbacks(self) -> Dict[str, PyTorchCallback]:
+        context = self.context
+        lm = self.lm
+        cb = PyTorchCallback()
+
+        def on_train_epoch_end(output: List[Any]) -> None:
+            lm.on_train_epoch_end(output)
+            lm.training_epoch_end(output)
+
+        def on_validation_epoch_end(outputs: List[Any]) -> None:
+            lm.on_validation_epoch_end()
+            lm.validation_epoch_end(outputs)
+
+        def on_train_epoch_start() -> None:
+            if context._current_batch_idx is not None:
+                type(lm).current_epoch = context.current_train_epoch()
+            lm.on_train_epoch_start()
+
+        def on_validation_epoch_start() -> None:
+            lm.on_validation_epoch_start()
+
+        cb.on_validation_epoch_start = on_validation_epoch_start
```

Comment:
Dynamically changing a member field of a Python object is generally a bad practice, which makes code hard to debug. Why don't you just define an inner class:

```
class PTLAdapterCallback(PyTorchCallback):
   def ...():
      ...
```